src/components/views/rooms/MessageComposerInput.js

```diff
@@ -1018,167 +1287,237 @@ export default class MessageComposerInput extends React.Component {
+                            room={this.props.room}
+                            shouldShowPillAvatar={shouldShowPillAvatar}
+                            isSelected={isSelected}
+                            />;
+                }
+                else if (Pill.isPillUrl(url)) {
+                    return <Pill
+                            url={url}
+                            room={this.props.room}
+                            shouldShowPillAvatar={shouldShowPillAvatar}
+                            isSelected={isSelected}
+                            />;
+                }
+                else {
+                    const { text } = node;
+                    return <a href={url} {...props.attributes}>
+                                { text }
+                           </a>;
+                }
+            }
+            case 'emoji': {
+                const { data } = node;
+                const emojiUnicode = data.get('emojiUnicode');
+                const uri = RichText.unicodeToEmojiUri(emojiUnicode);
+                const shortname = toShort(emojiUnicode);
+                const className = classNames('mx_emojione', {
+                    mx_emojione_selected: isSelected
+                });
+                return <img className={ className } src={ uri } title={ shortname } alt={ emojiUnicode }/>;
+            }
+        }
+    };
+
+    renderMark = props => {
+        const { children, mark, attributes } = props;
+        switch (mark.type) {
+            case 'bold':
+                return <strong {...attributes}>{children}</strong>;
+            case 'italic':
+                return <em {...attributes}>{children}</em>;
+            case 'code':
+                return <code {...attributes}>{children}</code>;
+            case 'underlined':
+                return <u {...attributes}>{children}</u>;
+            case 'deleted':
+                return <del {...attributes}>{children}</del>;
+        }
+    };
+
+    onFormatButtonClicked = (name, e) => {
         e.preventDefault(); // don't steal focus from the editor!
-        const command = {
-                code: 'code-block',
-                quote: 'blockquote',
-                bullet: 'unordered-list-item',
-                numbullet: 'ordered-list-item',
-            }[name] || name;
-        this.handleKeyCommand(command);
-    }
 
-    /* returns inline style and block type of current SelectionState so MessageComposer can render formatting
-     buttons. */
-    getSelectionInfo(editorState: EditorState) {
-        const styleName = {
-            BOLD: _td('bold'),
-            ITALIC: _td('italic'),
-            STRIKETHROUGH: _td('strike'),
-            UNDERLINE: _td('underline'),
-        };
+        // XXX: horrible evil hack to ensure the editor is focused so the act
+        // of focusing it doesn't then cancel the format button being pressed
+        if (document.activeElement && document.activeElement.className !== 'mx_MessageComposer_editor') {
```

Comment:
It seems odd to only call `handleKeyCommand` if something other than the composer was focused at this point. I think I'll ask about this IRL.

Reply:
have put this in as a fixme for now